src/System.Net.WebHeaderCollection/src/System/Net/HeaderInfoTable.cs

```diff
@@ -3,88 +3,146 @@
 
 using System.Collections;
+using System.Collections.Generic;
 using System.Collections.Specialized;
 
 namespace System.Net
 {
     internal class HeaderInfoTable
     {
-        private static Hashtable HeaderHashTable;
-        private static HeaderInfo UnknownHeaderInfo = new HeaderInfo(string.Empty, false, false, false);
+        private static readonly Func<string, string[]> s_singleParser = value => new[] { value };
+        private static readonly Func<string, string[]> s_multiParser = value => ParseValueHelper(value, isSetCookie: false);
+        private static readonly Func<string, string[]> s_setCookieParser = value => ParseValueHelper(value, isSetCookie: true);
+        private static readonly HeaderInfo s_unknownHeaderInfo = new HeaderInfo(string.Empty, false, false, false, s_singleParser);
+        private static readonly Hashtable s_headerHashTable;
+
+        private static string[] ParseValueHelper(string value, bool isSetCookie)
+        {
+            List<string> tempStringCollection = new List<string>();
+
+            bool inquote = false;
+            int chIndex = 0;
+            char[] vp = new char[value.Length];
+            string singleValue;
+
+            for (int i = 0; i < value.Length; i++)
+            {
+                if (value[i] == '\"')
+                {
+                    inquote = !inquote;
+                }
+                else if ((value[i] == ',') && !inquote)
+                {
+                    singleValue = new string(vp, 0, chIndex);
+                    if (!isSetCookie || !IsDuringExpiresAttributeParsing(singleValue))
+                    {
+                        tempStringCollection.Add(singleValue.Trim());
+                        chIndex = 0;
+                        continue;
+                    }
+                }
+                vp[chIndex++] = value[i];
+            }
+
+            // Now add the last of the header values to the stringtable.
+            if (chIndex != 0) {
+                singleValue = new string(vp, 0, chIndex);
+                tempStringCollection.Add(singleValue.Trim());
+            }
+
+            string[] stringArray = tempStringCollection.ToArray();
+            return stringArray;
+        }
+
+        // This method is to check if we are in the middle of parsing the Expires attribute
+        // for Set-Cookie header. It needs to check two conditions: 1. If current attribute
+        // is Expires. 2. Have we finished parsing it yet. Because the Expires attribute
+        // will contain exactly one comma, no comma means we are still parsing it.
+        private static bool IsDuringExpiresAttributeParsing(string singleValue)
+        {
+            string[] attributeArray = singleValue.Split(';');
+            string lastElement = attributeArray[attributeArray.Length - 1].Trim();
+            bool noComma = lastElement.IndexOf(',') < 0;
+
+            string lastAttribute = lastElement.Split('=')[0].Trim();
```

Comment:
What if lastElement didn't contain any equal signs?

Reply:
This function will be executed only when 1. It's SetCookie attribute 2. parsing multi-value header, and a comma has been detected for current single value.

According to RFC:
>  Each cookie begins with a name-value-pair, followed by zero or more attribute-value pairs.

For both cookie itself (name-value-pair) and its attributes (attribute-value pairs), they are all key-value pair follows this syntax: A = B. Therefore, lastElement will contain '=' sign.